simulator/src/delegate.rs

```diff
@@ -43,6 +53,32 @@ impl Delegate {
     #[inline(always)]
     #[tracing::instrument(skip_all)]
     pub fn recv_transaction_v0(&mut self, tx: &Transaction) -> Result<V0Response, Error> {
-        transaction_v0(tx, self.keypair, &mut self.db)
+        let target_endpoint = format!("{}/v0/rpc", self.target);
+        let request = Request::V0(V0Request::Transaction(tx.clone()));
+
+        let response = self
+            .client
+            .post(&target_endpoint)
+            .json(&request)
+            .send()
+            .map_err(|err| Error::ServerError {
+                reason: err.to_string(),
+            })?;
+
+        let response_text = response.text().map_err(|err| Error::ServerError {
+            reason: err.to_string(),
+        })?;
+        let v0_response: V0Response = serde_json::from_str(&response_text).map_err(|_| {
+            if response_text.contains("Atom has already been spent") {
```

Comment:
What is the best way to check for errors here?

Reply:
IIRC you can deserialize the error with the correct type using serde_json

Reply:
The api returns the error converted to a string. Am I able to get the string back into the structured type?
```rust
    match request {
        Request::V0(V0Request::Transaction(t)) => {
            let mut db = database.lock().unwrap();

            match transaction_v0(&t, keypair, &mut db) {
                Ok(response) => Json(Response::V0(response)).into_response(),
                Err(e) => Json(json!({ "error": e.to_string() })).into_response(),
            }
        }
    }
```

Sample error response as string: `{"error":"Atom has already been spent: 7a4abcbc805ed8a18759941fddd3d901a608aa124cd29b76efe8b6181c01c859"}` (just realized the proper signature was there all along)

I am able to properly deserialize the error if I modify the node to return the full structured error as json, like so
```rust
    match request {
        Request::V0(V0Request::Transaction(t)) => {
            let mut db = database.lock().unwrap();

            match transaction_v0(&t, keypair, &mut db) {
                Ok(response) => Json(Response::V0(response)).into_response(),
                // Err(e) => Json(json!({ "error": e.to_string() })).into_response(),
                Err(e) => Json(json!(e)).into_response()
            }
        }
    }
```
Which greatly simplifies the error handling here
```rust
      match serde_json::from_str::<V0Response>(&response_text) {
          Ok(v0_response) => Ok(v0_response),
          Err(_) => match serde_json::from_str::<Error>(&response_text) {
              Ok(api_error) => Err(api_error),
              Err(_) => Err(Error::ServerError {
                  reason: response_text,
              }),
          },
      }
```

Reply:
Yeah, I think that is the way to go. We leak some information on the errors but we can change it later if necessary.